### ISSUE TITLE
Crnk should respect JsonInclude [regression]

### ISSUE THREAD

**Issue by remmeier** [crnk-project/crnk-framework#25]
orignally raised by @ramblurr 

This is a regression from pre-2.8.

When the serialization was refactored, support for @JsonInclude was unintentionally dropped.

Before katharsis passed the resource objects to jackson directly, so the annotations were respected.

Now katharsis iterates over the properties and serializes them individually.

In our API we don't want to show null values for properties, so we have @JsonInclude(Include.NON_NULL) on many resources.

Implementing this would require

parsing the JsonInclude annotation in io.katharsis.core.internal.resource.AnnotationResourceInformationBuilder
storing it in the ResourceInformation
checking for it in io.katharsis.core.internal.resource.ResourceMapper#setAttributes
We actually lost quite a few nice jackson features when serialization was re-written for 2.8, among them:

@JsonInclude
@JsonAutoDetect
@JsonPropertyOrder

**Comment by remmeier** [comment]
@jsoninclude: ResourceField should have a similar property. AnnotationResourceInformationBuilder would parse the annotation accordingly and set the ResourceField property. Than the crnk implementation could account for it.

@JsonAutoDetect, @jsonpropertyorder: can be considered by AnnotationResourceInformationBuilder

**Comment by Ramblurr** [comment]
@remmeier grepping through crnk's codebase I see some JsonInclude usage.. is this now fixed?

**Comment by remmeier** [comment]
no, no so far, http://www.crnk.io/documentation/#_jackson_annotations

from a discoverability perspective I have to say I'm not so sure whether I should be overly fond of @JsonInclude. but it should be supported.

**Comment by remmeier** [comment]
not too much, I was experimenting a bit whether to pass all crnk-related serialization calls in ResourceMapper through a Jackson serializer. But not overly fond of what I got (
https://github.com/crnk-project/crnk-framework/tree/jackson_int). Maybe better to replicate the Jackson functionality without going much into Jackson.

**Comment by datagitlies** [comment]
I'm needing to use `@JsonInclude` and thus I'm affected by this issue.

Question: @remmeier in the original issue description @Ramblurr laid out what appears to be a pretty straightforward resolution:
>Implementing this would require
>
>parsing the JsonInclude annotation in
>io.katharsis.core.internal.resource.AnnotationResourceInformationBuilder
>storing it in the ResourceInformation
>checking for it in io.katharsis.core.internal.resource.ResourceMapper#setAttributes

Is that approach viable? AKA, should someone volunteer their time and open a pull request with the changes?

**Comment by a-dudko** [comment]
@remmeier is this issue relevant and going to be resolved?
@JsonInclude availability would be veeeeery useful.

**Comment by remmeier** [comment]
no not so far, but contributions are welcome.

A first step could also be to have a filter/interceptor within the DocumentMapper that allows to modify the JSON API request/response documents.

**Comment by a-dudko** [comment]
Should this be an internal crnk change or can such interceptor be manually introduced in project using crnk?

**Comment by vicmosin** [comment]
@remmeier not sure which filter/interceptor within the DocumentMapper you mean.. 
If we are talking about the solution - would it make sense to introduce a new enum to ResourceField (which will be actually filled out on startup) interface which actually will copy the values from json-include?

**Comment by remmeier** [comment]
a (new) filter/interceptor within DocumentMapper could provide the necessary hook to implement the workaround. 

Yes, extendending ResourceField for the target solution would be perfect. Then DocumentMapper performs its work accordingly. Similar to Jackson we could offer some built-in strategies for null, empty or always writting the attribute. But a developer should also be able to provide its custom strategy.

JacksonResourceFieldInformationProvider could then contribute to ResourceField to support the existing jackson annotation.

**Comment by jayantm** [comment]
Can anybody give some example to solve this?

**Comment by remmeier** [comment]
solve it or workaround it?

**Comment by jayantm** [comment]
ya even if its workaround thats fine.
